Scaffolding for .NET Isolated Analyzer

allantargino opened the pull request:
This PR scaffolds two projects:
- A .NET Standard project for Roslyn Analyzers targeting .NET Isolated analysis
- A .NET 6 xUnit Test project (to unit test the previous one)
  - All `Verifier` like classes were generated using "Analyzer with Code Fix" [project template](https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix#build-unit-tests). They are helpful when writing tests against analyzers.

### Pull request checklist

* [x] My changes **do not** require documentation changes
* [x] My changes **should not** be added to the release notes for the next release
* [x] My changes **do not** need to be backported to a previous version
* [x] I have added all required tests (Unit tests, E2E tests)
* [x] My changes **do not** require any extra work to be leveraged by OutOfProc SDKs
* [x] My changes **do not** change the version of the WebJobs.Extensions.DurableTask package
* [x] My changes **do not** add EventIds to our EventSource logs
* [x] My changes **should** be added to v3.x branch.

davidmrdavid commented on test/Worker.Extensions.DurableTask.Analyzers.Test/Worker.Extensions.DurableTask.Analyzers.Test.csproj:
```diff
@@ -0,0 +1,37 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+	<PropertyGroup>
+		<TargetFramework>net6.0</TargetFramework>
```
seems the indentation here is a bit large - can we make it more conservative? 2 to 4 spaces should suffice, or just a single tab

davidmrdavid commented on test/Worker.Extensions.DurableTask.Analyzers.Test/Verifiers/CSharpVerifierHelper.cs:
```diff
@@ -0,0 +1,33 @@
+﻿using Microsoft.CodeAnalysis;
```
let's add the copyright notice :)

```suggestion
﻿// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See LICENSE in the project root for license information.

﻿using Microsoft.CodeAnalysis;
```

davidmrdavid commented on src/Worker.Extensions.DurableTask.Analyzers/Worker.Extensions.DurableTask.Analyzers.csproj:
```diff
@@ -0,0 +1,30 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+	<PropertyGroup>
```
as noted elsewhere, the indentation is a bit large - can it be more conservative?

allantargino replied:
Good catch haha definitely!
I used VS to format it, but for some reason it used tabs over spaces, I am going to fix it. Thanks

davidmrdavid commented on test/Worker.Extensions.DurableTask.Analyzers.Test/Verifiers/CSharpAnalyzerVerifier`1+Test.cs:
```diff
@@ -0,0 +1,26 @@
+﻿using Microsoft.CodeAnalysis.CSharp.Testing;
```
adding copyright notice :) 
```suggestion
﻿// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See LICENSE in the project root for license information.

﻿using Microsoft.CodeAnalysis.CSharp.Testing;
```

davidmrdavid commented on test/Worker.Extensions.DurableTask.Analyzers.Test/Verifiers/CSharpAnalyzerVerifier`1.cs:
```diff
@@ -0,0 +1,36 @@
+﻿using Microsoft.CodeAnalysis;
```
adding copyright notice :)

```suggestion
﻿// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See LICENSE in the project root for license information.

﻿using Microsoft.CodeAnalysis;
```

davidmrdavid commented on test/Worker.Extensions.DurableTask.Analyzers.Test/Verifiers/CSharpCodeFixVerifier`2+Test.cs:
```diff
@@ -0,0 +1,28 @@
+﻿using Microsoft.CodeAnalysis.CodeFixes;
```
```suggestion
﻿// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See LICENSE in the project root for license information.

﻿using Microsoft.CodeAnalysis.CodeFixes;
```

davidmrdavid commented on test/Worker.Extensions.DurableTask.Analyzers.Test/Verifiers/CSharpCodeFixVerifier`2.cs:
```diff
@@ -0,0 +1,59 @@
+﻿using Microsoft.CodeAnalysis;
```
```suggestion
﻿// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the MIT License. See LICENSE in the project root for license information.

﻿using Microsoft.CodeAnalysis;
```

davidmrdavid reviewed (commented):
Some minor comments, mostly around indentation and copyright notice.

ALso - I notic3ed some files have backticks (i.e " ` " ) in their name, such as "CsharpAnalyzerVerifier`1.cs". Was that intentional? If not, I'd prefer to remove the backticks to make the file names easily searchable :-)

bachuv commented on src/Worker.Extensions.DurableTask.Analyzers/Resources.Designer.cs:
```diff
@@ -0,0 +1,63 @@
+    // class via a tool like ResGen or Visual Studio.
+    // To add or remove a member, edit your .ResX file then rerun ResGen
+    // with the /str option, or rebuild your VS project.
+    [global::System.CodeDom.Compiler.GeneratedCodeAttribute("System.Resources.Tools.StronglyTypedResourceBuilder", "17.0.0.0")]
+    [global::System.Diagnostics.DebuggerNonUserCodeAttribute()]
+    [global::System.Runtime.CompilerServices.CompilerGeneratedAttribute()]
```
Is this indentation expected?

allantargino replied:
This file is auto generated - even if I change it I believe it will be regenerated like that :(

bachuv commented on test/Worker.Extensions.DurableTask.Analyzers.Test/Verifiers/CSharpVerifierHelper.cs:
```diff
@@ -0,0 +1,33 @@
+        /// </summary>
+        internal static ImmutableDictionary<string, ReportDiagnostic> NullableWarnings { get; } = GetNullableWarningsFromCompiler();
+
+        private static ImmutableDictionary<string, ReportDiagnostic> GetNullableWarningsFromCompiler()
```
How is the `CSharpVerifierHelper` used in our scenario with the .NET Isolated Analyzer? Do we want to make all nullable reference types diagnostics an error instead of a warning? Do we need this verifier helper or can we remove it?

allantargino replied:
Answered below (along with the other comments :))

bachuv replied:
Thanks! How are we using this file? Is there any reason that we need to make nullable reference types diagnostics an error instead of a warning?

allantargino replied:
This was part of the code generated, I assuming they would like to simulate how the compiler used by final users behave. For our case, we don't need that behavior (because we have full control of the written tests and we aren't looking for this kind of compilation errors). So, I am going to remove this additional configuration. If needed later, we can enable it again.

bachuv commented on test/Worker.Extensions.DurableTask.Analyzers.Test/Verifiers/CSharpAnalyzerVerifier`1.cs:
Can you add a comment at the beginning of the file explaining this verifier and and then another comment in verifier 2? It will be helpful to understand the high level purpose and give other team members context.

bachuv commented on test/Worker.Extensions.DurableTask.Analyzers.Test/Verifiers/CSharpAnalyzerVerifier`1.cs:
Can we update the file names for the verifiers to more accurately reflect what they are doing? Looks like one provides a code fix, but the other doesn't.

bachuv reviewed (commented):
Thanks for opening this PR!

allantargino commented:
Hi @davidmrdavid and @bachuv, thanks for the review!
I just fixed the indentation, added copyrights and removed the ` from the file names. They were actually generated by VS template, so I didn't change their names or content. But I do agree the names shouldn't have that special char!

allantargino commented:
@bachuv, those files you commented are part of a set of helpers provided by Roslyn team to write clean and simple unit tests for Analyzers, they are generated as part of a new Analyzer project. I am sorry it seems a bit disconnected and random files, I should have given more context, my bad :)

For instance, this is how a unit test is written using those built-in utilities (which is pretty close to what I am going to send in my next PR):

```cs
using SomeNamespace.Analyzer;
using VerifyCS = Worker.Extensions.DurableTask.Analyzers.Tests.CSharpAnalyzerVerifier<SomeNamespace.Analyzer.SomeAnalyzer>;

namespace SomeNamespace.Tests;

public class UnitTest1
{
    [Fact]
    public async Task Test1()
    {
        var test = @"";

        await VerifyCS.VerifyAnalyzerAsync(test);
    }

    [Fact]
    public async Task Test2()
    {
        var test = @"
using System;

class Someclass
{   
    static void Write()
    {
        var now1 = {|#0:DateTime.Now|};
        Console.WriteLine(now1);
    }
}";

        var expected = VerifyCS
            .Diagnostic(SomeAnalyzer.DiagnosticId)
            .WithLocation(0)
            .WithArguments("System.DateTime.Now");

        await VerifyCS.VerifyAnalyzerAsync(test, expected);
    }
}
```

So, in general terms, the files do:
- `CSharpAnalyzerVerifier.cs`: verifies (runs tests against) a CSharp [DiagnosticAnalyzer](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnostics.diagnosticanalyzer?view=roslyn-dotnet-4.7.0)
- `CSharpAnalyzerVerifier.Test.cs` has the "Test" model for a verification of a CSharp [DiagnosticAnalyzer](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnostics.diagnosticanalyzer?view=roslyn-dotnet-4.7.0)
- `CSharpCodeFixVerifier.cs` verifies a CSharp [DiagnosticAnalyzer](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnostics.diagnosticanalyzer?view=roslyn-dotnet-4.7.0) + its related CSharp [CodeFixProvider](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.codefixes.codefixprovider?view=roslyn-dotnet-4.7.0)
- `CSharpCodeFixVerifier.Test.cs` has the "Test" model for a verification of a CSharp [DiagnosticAnalyzer](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnostics.diagnosticanalyzer?view=roslyn-dotnet-4.7.0) + its related CSharp [CodeFixProvider](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.codefixes.codefixprovider?view=roslyn-dotnet-4.7.0)

Giving this context, specially that those files are generated during scaffolding, should we change their names? IMHO, I believe we should keep their names as close as possible to the generated files, so that when the Roslyn team updates them, we can just replace our ones.

Thoughts?

bachuv commented:
@allantargino Thanks for the in-depth explanation! Can you add this information in a README file for the project?

In terms of the file names, I agree that we can leave the names if they are auto generated. Can you take a look at how those files are structured in the existing analyzer project for In Proc though?

allantargino commented:
I just simplified the files' structure, to:
- remove the not used nullability compiler configuration (as per discussion above)
- align naming to the previous analyzer testing project:
  ![image](https://github.com/Azure/azure-functions-durable-extension/assets/13934447/25d2b6bb-4a6b-41ba-b585-ddeb4d5c8c34)

Besides, I added the information above to a simple README file (to be expanded later).

davidmrdavid reviewed (approved):
LGTM but please get @bachuv's approval.
Also unsure if this matches @jviau's suggestion from today regarding project structure and naming, but I believe you can address that in a follow-up PR

allantargino commented:
Thanks for the approval, @davidmrdavid!
As per yesterday's internal discussion, we are going to develop the analyzer in the `microsoft/durabletask-dotnet` repository. The analyzer should be able to analyze not only Durable Functions code, but also usage in DTFx (although initially only focusing on matching parity with the current DF in-proc analyzer).

I am going to close this PR and move it to https://github.com/microsoft/durabletask-dotnet/pull/283
Thanks!
